compiler/src/middle_end/matchcomp.re

```diff
@@ -23,6 +22,13 @@ type decision_tree =
     Leaf(
       int,
     )
+  | /** Represents a guarded branch. The left tree corresponds to a successful
+      guard check, and the right tree corresponds to a failed guard check. */
+    Guard(
+      match_branch,
+      decision_tree,
+      decision_tree,
+    )
```

Comment:
This was simpler, but potentially this could/should be `pattern, expression, decision_tree, decision_tree` to guarantee that the `match_branch` actually has a guard, but I feel it's a non-issue since it's contained within this module.